### PR TITLE
Gitlab CI: Windows Configs

### LINKED ISSUE #44163: 'spack install' on Windows errors with DLL trying to link to itself
### Steps to reproduce

spack uninstall --all
spack install paraview

### Error message

[spack] C:\Users\dan.lipsa\projects\spack>spack install paraview
==> Installing bzip2-1.0.8-lua5n47dgpejpbfinuqewbceq3yoaca3 [1/38]
==> No binary for bzip2-1.0.8-lua5n47dgpejpbfinuqewbceq3yoaca3 found: installing from source
==> Using cached archive: C:\Users\dan.lipsa\projects\spack\var\spack\cache\_source-cache\archive\ab\ab5a03176ee106d3f0fa90e381da478ddae405918153cca248e682cd0c4a2269.tar.gz
==> Ran patch() for bzip2
==> bzip2: Successfully installed bzip2-1.0.8-lua5n47dgpejpbfinuqewbceq3yoaca3
  Stage: 0.10s.  Install: 6.62s.  Post-install: 0.14s.  Total: 6.90s
[+] C:\Users\dan.lipsa\projects\spack\opt\spack\windows-windows10.0.22631-skylake\msvc-19.39.33523\bzip2-1.0.8-lua5n47dgpejpbfinuqewbceq3yoaca3
[+] C:\Program Files\cmake (external cmake-3.29.0-rc1-fj44ws7t4rj44farqvfogb2nqewueexs)
[+] C:\Program Files\Microsoft Visual Studio\2022\Professional\Common7\IDE\CommonExtensions\Microsoft\CMake\Ninja (external ninja-1.11.0-3mpj42ci3xgxgmwygavomw57fsc3ipir)
==> Error: [WinError 5] Access is denied: 'C:\\Program Files (x86)\\Windows Kits\\10\\bin\\10.0.17763.0\\x86\\Microsoft.WindowsAzure.Storage.dll' -> 'C:\\Program Files (x86)\\Windows Kits\\10\\bin\\Microsoft.WindowsAzure.Storage.dll'

### Information on your system

* **Spack:** 0.22.0.dev0 (125206d44d6ecf167624531dff3631b126791849)
* **Python:** 3.11.9
* **Platform:** windows-windows10.0.22631-skylake
* **Concretizer:** clingo

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by johnwparent**
Support for Gitlab CI on Windows is here! 
This PR adds the config changes required to configure and execute Gitlab pipelines running Windows builds on Windows runners using the existing Gitlab CI infrastructure (and newly added Windows side infra). 

Adds support for generating child pipelines dispatched to Windows runners
Refactors the relevant pre-scripts, scripts, and post scripts to be compatible with Windows
Adds Windows config section describing Windows jobs
Adds VTK as Windows build stack (to be expanded later).

**DO NOT MERGE THIS** without validating pre-reqs have been met with a member of the Gitlab CI team and/or @johnwparent, @scottwittenburg, @kwryankrattiger, or @mvandenburgh.
PR Prereqs:
- [x] https://github.com/spack/spack/pull/43965
- [x] https://github.com/spack/spack/pull/43457
- [x] https://github.com/spack/spack/pull/43851

Manual Checks:
- [x] Validate on protected pipelines
- [x] Deploy Windows Spack Infrastructure for non staging Gitlab instances.

### CODE REVIEW

**Comment by johnwparent** [comment]
CC @scottwittenburg @kwryankrattiger @zackgalbreath

File: share/spack/gitlab/cloud_pipelines/configs/ci.yaml [hunk 1]
```diff
@@ -37,9 +37,10 @@ ci:
         CI_JOB_SIZE: "default"
         CI_GPG_KEY_ROOT: /mnt/key
         # SPACK_VERBOSE_SCRIPT: "1"
-      id_tokens:
-        GITLAB_OIDC_TOKEN:
-          aud: "${OIDC_TOKEN_AUDIENCE}"
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/configs/ci.yaml]: I don't think this block should be removed, it's how build jobs get permission to write to the appropriate bucket/prefix.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 2]
```diff
@@ -1,6 +1,7 @@
 stages: [ "generate", "build", "publish" ]
 
 variables:
+  SPACK_CI_ENABLE_STACKS: /^(windows-vis)$/
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: This bit has to go before merging, and probably before any final testing in staging as well.  Since this PR affects so much non-windows functionality, it will be good to make sure everything else is still building.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 3]
```diff
@@ -111,7 +119,8 @@ default:
         SPACK_PIPELINE_TYPE: "spack_pull_request"
         # TODO: We can remove this when we drop the "deprecated" stack
         PUSH_BUILDCACHE_DEPRECATED: "${PR_MIRROR_PUSH_DOMAIN}/${CI_COMMIT_REF_NAME}/${SPACK_CI_STACK_NAME}"
-        SPACK_PRUNE_UNTOUCHED: "True"
+        # SPACK_PRUNE_UNTOUCHED: "True"
+        SPACK_PRUNE_UNTOUCHED: "False"
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: These changes need undone before merging.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 4]
```diff
@@ -236,7 +238,7 @@ default:
     max: 2
     when:
       - always
-  tags: ["spack", "public", "medium", "x86_64"]
+  tags: ["spack", "public", "medium", "x86_64", "norun"]
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: I'm not sure what this `norun` tag is, should it still be here?

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 5]
```diff
@@ -434,8 +436,8 @@ e4s-rocm-external-build:
     SPACK_CI_STACK_NAME: e4s-oneapi
 
 e4s-oneapi-generate:
-  extends: [ ".e4s-oneapi", ".generate-x86_64"]
-  image:  ghcr.io/spack/ubuntu22.04-runner-amd64-oneapi-2024.0.0:2024.01.16b
+  extends: [ ".e4s-oneapi", ".generate"]
+  image: ecpe4s/ubuntu20.04-runner-x86_64-oneapi:2023.07.21
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Let's not change the image tags for other stacks?  Probably an artifact of the rebase, but needs to be fixed.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 6]
```diff
@@ -434,8 +436,8 @@ e4s-rocm-external-build:
     SPACK_CI_STACK_NAME: e4s-oneapi
 
 e4s-oneapi-generate:
-  extends: [ ".e4s-oneapi", ".generate-x86_64"]
-  image:  ghcr.io/spack/ubuntu22.04-runner-amd64-oneapi-2024.0.0:2024.01.16b
+  extends: [ ".e4s-oneapi", ".generate"]
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Did you mean to change this oneapi generate definition?

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 7]
```diff
@@ -452,7 +454,7 @@ e4s-oneapi-build:
 # E4S on Power
 ########################################
 .e4s-power-generate-tags-and-image:
-  image: { "name": "ghcr.io/spack/ubuntu20.04-runner-ppc64-gcc-11.4:2023.08.01", "entrypoint": [""] }
+  image: { "name": "ecpe4s/ubuntu20.04-runner-ppc64le:2023-01-01", "entrypoint": [""] }
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Unrelated image change should be fixed.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 8]
```diff
@@ -461,7 +463,7 @@ e4s-oneapi-build:
     SPACK_CI_STACK_NAME: e4s-power
 
 e4s-power-generate:
-  extends: [ ".e4s-power", ".generate-x86_64", ".e4s-power-generate-tags-and-image"]
+  extends: [ ".e4s-power", ".generate", ".e4s-power-generate-tags-and-image"]
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: I feel like this probably still needs to extend the arch-specific generate definition

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 9]
```diff
@@ -483,7 +485,7 @@ e4s-power-build:
     SPACK_CI_STACK_NAME: build_systems
 
 build_systems-generate:
-  extends: [ ".build_systems", ".generate-x86_64"]
+  extends: [ ".build_systems", ".generate"]
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate definition

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 10]
```diff
@@ -496,28 +498,6 @@ build_systems-build:
...
-  extends: [ ".developer-tools", ".build" ]
-  trigger:
-    include:
-      - artifact: jobs_scratch_dir/cloud-ci-pipeline.yml
-        job: developer-tools-generate
-    strategy: depend
-  needs:
-    - artifacts: True
-      job: developer-tools-generate
-
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: I don't think the `developer-tools` stack should be removed :wink:

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 11]
```diff
@@ -551,7 +531,7 @@ developer-tools-manylinux2014-build:
     SPACK_CI_STACK_NAME: radiuss
 
 radiuss-generate:
-  extends: [ ".radiuss", ".generate-x86_64" ]
+  extends: [ ".radiuss", ".generate" ]
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate thing.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 12]
```diff
@@ -583,7 +563,7 @@ radiuss-build:
     SPACK_CI_STACK_NAME: radiuss-aws
 
 radiuss-aws-generate:
-  extends: [ ".radiuss-aws", ".generate-x86_64", ".radiuss-aws-overrides", ".tags-x86_64_v4" ]
+  extends: [ ".radiuss-aws", ".generate", ".radiuss-aws-overrides", ".tags-x86_64_v4" ]
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate thing.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 13]
```diff
@@ -627,8 +607,8 @@ radiuss-aws-aarch64-build:
     SPACK_CI_STACK_NAME: data-vis-sdk
 
 data-vis-sdk-generate:
-  extends: [ ".data-vis-sdk", ".generate-x86_64"]
-  image: ghcr.io/spack/ubuntu20.04-runner-x86_64:2023-01-01
+  extends: [ ".data-vis-sdk", ".generate"]
+  image: ecpe4s/ubuntu20.04-runner-x86_64:2023-01-01
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate thing. And don't change the image back to dockerhub please.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 14]
```diff
@@ -657,7 +637,7 @@ data-vis-sdk-build:
     SPACK_CI_STACK_NAME: aws-isc
 
 aws-isc-generate:
-  extends: [ ".aws-isc", ".generate-x86_64", ".aws-isc-overrides", ".tags-x86_64_v4" ]
+  extends: [ ".aws-isc", ".generate", ".aws-isc-overrides", ".tags-x86_64_v4" ]
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate thing.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 15]
```diff
@@ -701,8 +681,8 @@ aws-isc-aarch64-build:
     SPACK_CI_STACK_NAME: tutorial
 
 tutorial-generate:
-  extends: [ ".tutorial", ".generate-x86_64"]
-  image: ghcr.io/spack/tutorial-ubuntu-22.04:v2023-10-30
+  extends: [ ".tutorial", ".generate"]
+  image: ghcr.io/spack/tutorial-ubuntu-22.04:v2023-05-07
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate thing.  And don't move us back to an older image please.

File: share/spack/gitlab/cloud_pipelines/scripts/common/expand_vars.py [hunk 16]
```diff
@@ -0,0 +1,10 @@
+import argparse
+import os
+
+parser = argparse.ArgumentParser()
+parser.add_argument("input", type=argparse.FileType("r"))
+parser.add_argument("out", type=argparse.FileType("w"))
+
+args = parser.parse_args()
+
+args.out.write(os.path.expandvars(args.intput.read()))
```
Comment (johnwparent) [inline share/spack/gitlab/cloud_pipelines/scripts/common/expand_vars.py]:
```suggestion
args.out.write(os.path.expandvars(args.input.read()))
```

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 17]
```diff
@@ -724,8 +704,8 @@ tutorial-build:
     SPACK_CI_STACK_NAME: ml-linux-x86_64-cpu
 
 ml-linux-x86_64-cpu-generate:
-  extends: [ ".generate-x86_64", .ml-linux-x86_64-cpu, ".tags-x86_64_v4" ]
-  image: ghcr.io/spack/linux-ubuntu22.04-x86_64_v2:v2024-01-29
+  extends: [ ".generate", .ml-linux-x86_64-cpu, ".tags-x86_64_v4" ]
+  image: ghcr.io/spack/linux-ubuntu22.04-x86_64_v2:nightly
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate thing.  And we don't want a nightly image here.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 18]
```diff
@@ -747,8 +727,8 @@ ml-linux-x86_64-cpu-build:
     SPACK_CI_STACK_NAME: ml-linux-x86_64-cuda
 
 ml-linux-x86_64-cuda-generate:
-  extends: [ ".generate-x86_64", .ml-linux-x86_64-cuda, ".tags-x86_64_v4" ]
-  image: ghcr.io/spack/linux-ubuntu22.04-x86_64_v2:v2024-01-29
+  extends: [ ".generate", .ml-linux-x86_64-cuda, ".tags-x86_64_v4" ]
+  image: ghcr.io/spack/linux-ubuntu22.04-x86_64_v2:nightly
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Ditto the arch-specific generate thing.  And the nightly image thing.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 19]
```diff
@@ -839,11 +819,6 @@ aws-pcluster-build-x86_64_v4:
     - artifacts: True
       job: aws-pcluster-generate-x86_64_v4
 
-# Neoverse_v1 (one pipeline per target)
-.aws-pcluster-neoverse_v1:
-  variables:
-    SPACK_CI_STACK_NAME: aws-pcluster-neoverse_v1
-
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: You probably didn't intend to remove this block?

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 20]
```diff
@@ -919,3 +894,36 @@ e4s-cray-sles-build:
...
+# aws-pcluster-build-neoverse_v1:
+#   extends: [  ".linux_neoverse_v1", ".aws-pcluster-neoverse_v1", ".build" ]
+#   trigger:
+#     include:
+#       - artifact: jobs_scratch_dir/cloud-ci-pipeline.yml
+#         job: aws-pcluster-generate-neoverse_v1
+#     strategy: depend
+#   needs:
+#     - artifacts: True
+#       job: aws-pcluster-generate-neoverse_v1
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Better double check on the current status of the pcluster jobs, I don't think they're commented out anymore, but I'm not sure.

File: share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml [hunk 21]
```diff
@@ -1,5 +1,11 @@
 spack:
   view: false
+
+  include:
+  - $spack/share/spack/gitlab/cloud_pipelines/configs/linux/x86_64_v3
+  - $spack/share/spack/gitlab/cloud_pipelines/configs/linux
+  - $spack/share/spack/gitlab/cloud_pipelines/configs
```
Comment (scottwittenburg) [inline share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml]: This is unexpected to me, is it related to windows ci or an artifact of the rebase or testing or something?

Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml]: @kwryankrattiger this is from commit https://github.com/spack/spack/pull/43967/commits/9f3f40c95ad89f440697f72c96efd15216b21619, I assume it's a refactor because the Windows side of configuration made the previous scope behavior no longer sufficiently generic?

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml (hunk 4 above)
Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: Title of the commit that added this was "Disable non-windows stacks using bad tags" LOL. Just going to drop this.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml (hunk 20 above)
Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: At one point I commented out the entire yaml except things required for Windows,  this is cruft.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 22]
```diff
@@ -124,49 +133,23 @@ default:
...
-    - spack
-      --config-scope "${SPACK_CI_CONFIG_ROOT}"
-      --config-scope "${SPACK_CI_CONFIG_ROOT}/${SPACK_TARGET_PLATFORM}"
-      --config-scope "${SPACK_CI_CONFIG_ROOT}/${SPACK_TARGET_PLATFORM}/${SPACK_TARGET_ARCH}"
-      ${CI_STACK_CONFIG_SCOPES}
-      audit configs
-    - spack python -c "import os,sys; print(os.path.expandvars(sys.stdin.read()))"
-      < "${SPACK_CI_CONFIG_ROOT}/${PIPELINE_MIRROR_TEMPLATE}" > "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
+      # - spack
+      #   audit configs
```
Comment (johnwparent) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: @johnwparent Make sure these invoked somewhere not commented and remove commented versions once confirmed

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 23]
```diff
@@ -124,49 +133,23 @@ default:
...
-    - mkdir -p "${CI_PROJECT_DIR}/jobs_scratch_dir"
-    - spack
-      --config-scope "${SPACK_CI_CONFIG_ROOT}"
-      --config-scope "${SPACK_CI_CONFIG_ROOT}/${SPACK_TARGET_PLATFORM}"
-      --config-scope "${SPACK_CI_CONFIG_ROOT}/${SPACK_TARGET_PLATFORM}/${SPACK_TARGET_ARCH}"
-      ${CI_STACK_CONFIG_SCOPES}
-      config blame > "${CI_PROJECT_DIR}/jobs_scratch_dir/spack.yaml.blame"
+    - mkdir "${CI_PROJECT_DIR}/jobs_scratch_dir"
+      # - spack
+      #   config blame > "${CI_PROJECT_DIR}/jobs_scratch_dir/spack.yaml.blame"
```
Comment (johnwparent) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml]: @johnwparent Ensure blame is run at some point

File: share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml (hunk 21 above)
Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml]: This will come in another PR someday. I think I did this because the windows ci PR was originally branched off of another refactor branch I was playing with.

Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml]: So revert all the include changes? Just want to be clear here.

Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml]: yeah, you should be able to drop all of those. This is redundant with the `-C` stuff in the `.gitlab-ci.yml`.

I will make a  point of making another PR to do this later, the current design is kind of unfortunate.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 24]
```diff
@@ -124,49 +131,23 @@ default:
...
+      # - spack
+      #   audit configs
+    - python ${SPACK_CI_SCRIPTS_ROOT}/common/expand_vars.py
+      "${SPACK_CI_CONFIG_ROOT}/${PIPELINE_MIRROR_TEMPLATE}"
+      "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
     - spack config add -f "${SPACK_CI_CONFIG_ROOT}/mirrors.yaml"
-    - mkdir -p "${CI_PROJECT_DIR}/jobs_scratch_dir"
-    - spack
-      --config-scope "${SPACK_CI_CONFIG_ROOT}"
-      --config-scope "${SPACK_CI_CONFIG_ROOT}/${SPACK_TARGET_PLATFORM}"
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:145]: `--config-scope` should still be used here. Otherwise we need to handle the scope includes in all of the `spack.yaml`s

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml [hunk 25]
```diff
@@ -204,6 +195,15 @@ default:
   extends: [ ".generate-base" ]
   tags: ["spack", "public", "medium", "neoverse_v2"]
 
+.generate-win64:
+  extends: [ ".base-job", ".generate-common" ]
+  before_script:
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:206]: I think this needs an empty after script?

File: share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml (hunk 21 above)
Comment (kwryankrattiger) [reply share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml]: Ugh, see other comment...

File: share/spack/gitlab/cloud_pipelines/stacks/windows-vis/spack.yaml [hunk 26]
```diff
@@ -0,0 +1,18 @@
+# Windows Visualization Stack
+# maintainers:
+# - John Parent (@johnwparent)
+# - Ryan Krattiger (@kwryankrattiger)
+
+spack:
+  view: false
+
+  include:
+  - $spack/share/spack/gitlab/cloud_pipelines/configs/win64/x86_64
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/windows-vis/spack.yaml]: I think getting `--config-scope` working is better for now.

File: share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml (hunk 25 above)
Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/.gitlab-ci.yml:206]: Added the equivalent for Windows

File: share/spack/gitlab/cloud_pipelines/stacks/windows-vis/spack.yaml (hunk 26 above)
Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/stacks/windows-vis/spack.yaml]: Done, removed these files and restored `--config-scope`

File: share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml (hunk 21 above)
Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/stacks/build_systems/spack.yaml]: By `-C` is assume you mean the `--config-scope` changes, in which case, addressed.

File: share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+
+      before_script::
+      - fsutil 8dot3name set C:\ 0
+      - . .\share\spack\setup-env.ps1
+      - spack.ps1 compiler find
+      - spack.ps1 external find --not-buildable  --path "C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake" --path "C:/Program Files/Microsoft Visual Studio/2022/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/Ninja" --path "C:/Program Files (x86)/Windows Kits/10/Lib/10.0.20348.0/um/x64" --path "C:/Program Files (x86)/Windows Kits/10/bin/10.0.20348.0/x64" cmake ninja win-sdk wgl
+
+      script::
+      - spack.ps1 env activate --without-view ${SPACK_CONCRETE_ENV_DIR}
+      - spack.ps1 config add "config:install_tree:projections:${SPACK_JOB_SPEC_PKG_NAME}:'morepadding/{architecture}/{compiler.name}-{compiler.version}/{name}-{version}-{hash}'"
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]: shorter path here?

File: share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml [hunk 28]
```diff
@@ -0,0 +1,20 @@
+ci:
+  pipeline-gen:
+  - build-job:
+      after_script::
+      - Write-Output "Done"
+
+      before_script::
+      - fsutil 8dot3name set C:\ 0
+      - . .\share\spack\setup-env.ps1
+      - spack.ps1 compiler find
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]: This should not be needed in the build-job, the generate job does this and caches the result in the `spack.yaml` we activate here for the environment

File: share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml (hunk 27 above)
Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]: Addressed in latest push

File: share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml (hunk 28 above)
Comment (johnwparent) [reply share/spack/gitlab/cloud_pipelines/configs/win64/ci.yaml]: Addressed in latest push.

**Review by kwryankrattiger — APPROVED** [review]

**Comment by adamjstewart** [comment]
Excited to see this merged!

I know we're still a long way off from getting all dependencies working, but I would be keen on porting our ML CI stacks to build on Windows someday. Let me know if there's anything I can do to help with the Python on Windows effort.

**Comment by johnwparent** [comment]
> Excited to see this merged!
> 
> I know we're still a long way off from getting all dependencies working, but I would be keen on porting our ML CI stacks to build on Windows someday. Let me know if there's anything I can do to help with the Python on Windows effort.

Porting all the stacks is the endgame of Spack on Windows, so I'm happy to support any effort doing so. It's not on my radar for the next month or so, but if you have specific packages/stacks you'd like to see, I'd be more than happy to iterate over slack, feel free to DM me.
